File: external/sigma/sigma.R
#!/usr/bin/env Rscript
library(SigMA)
library(optparse)

# ------------------- Main Analysis -------------------
runSigMA <- function(inputdir,
                    outputfile = "sigma-results.csv",
                    file_type = "maf",
                    tumor_type = "breast",
                    data = "seqcap",
                    do_assign = TRUE,
                    do_mva = TRUE,
                    lite_format = TRUE,
                    check_msi = TRUE
                    ) {
  genomes_matrix <- make_matrix(inputdir, file_type = file_type)
  genomes <- conv_snv_matrix_to_df(genomes_matrix)
  genome_file <- gsub(".maf", "_maf.csv", inputdir)
  write.table(genomes,
              genome_file,
              sep = ",",
              row.names = FALSE,
              col.names = TRUE,
              quote = FALSE)

  run(genome_file,
      output_file = outputfile,
      tumor_type = tumor_type,
      data = data,
      do_assign = do_assign,
      do_mva = do_mva,
      lite_format = lite_format,
      check_msi = check_msi)
  file.remove(genome_file)
}


option_list <- list(
  make_option(c("-i", "--input_file"), type = "character", default = NULL,
              help = "pointer to the directory where input vcf or maf files reside,example './sample.maf' ",
              metavar = "character"),
  make_option(c("-o", "--output_file"), type = "character", default = "sigma-results.csv",
              help = "the name of output file [default= %default]",
              metavar = "character"),
  make_option(c("-f", "--file_type"), type = "character", default = "maf",
              help = " 'maf' or 'vcf' [default= %default]", metavar = "character"),
  make_option(c("-t", "--tumor_type"), type = "character", default = "breast",
              help = "the options are 'bladder', 'bone_other' (Ewing's sarcoma or Chordoma), 'breast',
                'crc','eso', 'gbm', 'lung', 'lymph', 'medullo', 'osteo', 'ovary', 'panc_ad', 'panc_en', 
                'prost', 'stomach', 'thy', or 'uterus'.[default= %default]",
              metavar = "character"),
  make_option(c("-d", "--data"), type = "character", default = "seqcap",
              help = "the options are 'msk' (for a panel that is similar size to MSK-Impact panel with 410 genes),
                'seqcap' (for whole exome sequencing), 'seqcap_probe' (64 Mb SeqCap EZ Probe v3), or 'wgs' 
                (for whole genome sequencing) [default= %default]",
              metavar = "character"),
  make_option(c("-a", "--do_assign"), type = "logical", default = TRUE,
              help = "a boolean for whether a cutoff should be applied to determine the final decision or just 
                the features should be returned [default= %default]"),
  make_option(c("-m", "--do_mva"), type = "logical", default = TRUE,
              help = "a boolean for whether multivariate analysis should be run [default= %default]"),
  make_option(c("-F", "--lite_format"), type = "logical", default = TRUE,
              help = "saves the output in a lite format when set to true [default= %default]"),
  make_option(c("-c", "--check_msi"), type = "logical", default = TRUE,
              help = "a boolean which determines whether the user wants to identify micro-sattelite instable tumors[default= %default]")
)

opt_parser <- OptionParser(option_list = option_list)
opt <- parse_args(opt_parser)

if (is.null(opt$input_file)) {
  print_help(opt_parser)
  stop("At least one argument must be supplied (input file).n", call. = FALSE)
}

print(sprintf("Options: %s = %s", names(opt), opt))

runSigMA(inputdir = opt$input_file,
         outputfile = opt$output_file,
         file_type = opt$file_type,
         tumor_type = opt$tumor_type,
         data = opt$data,
         do_assign = opt$do_assign,
         do_mva = opt$do_mva,
         lite_format = opt$lite_format,
         check_msi = opt$check_msi)

File: plugins/tools/sigma.clj
(ns plugins.tools.sigma
  (:require [clojure.core.async :as async]
            [clojure.data.json :as json]
            [clojure.spec.alpha :as s]
            [clojure.tools.logging :as log]
            [spec-tools.core :as st]
            [clojure.string :as str]
            [spec-tools.json-schema :as json-schema]
            [tservice.config :refer [get-workdir]]
            [tservice.events :as events]
            [plugins.wrappers.sigma :as sigma-lib]
            [tservice.lib.fs :as fs-lib]
            [tservice.util :as u]))

;;; ------------------------------------------------ Event Specs ------------------------------------------------
(s/def ::filetype
  (st/spec
   {:spec                #(#{"maf" "vcf"} %)
    :type                :string
    :description         "File Type (maf or vcf)"
    :swagger/default     "vcf"
    :reason              "Only support maf or vcf"}))

(s/def ::tumor_type
  (st/spec
   {:spec                #(#{"bladder" "bone_other" "breast" "crc" "eso" "gbm" "lung" "lymph" "medullo" "osteo" "ovary" "panc_ad" "panc_en" "prost" "stomach" "thy" "uterus"} %)
    :type                :string
    :description         "The options are 'bladder', 'bone_other' (Ewing's sarcoma or Chordoma), 'breast', 'crc','eso', 'gbm', 'lung', 'lymph', 'medullo', 'osteo', 'ovary', 'panc_ad', 'panc_en', 'prost', 'stomach', 'thy', or 'uterus'."
    :swagger/default     "breast"
    :reason              "Only support bladder, bone_other, breast, crc, eso, gbm, lung, lymph, medullo, osteo, ovary, panc_ad, panc_en, prost, stomach, thy or uterus"}))

(s/def ::data
  (st/spec
   {:spec                #(#{"msk" "seqcap" "seqcap_probe" "wgs"} %)
    :type                :string
    :description         "The options are 'msk' (for a panel that is similar size to MSK-Impact panel with 410 genes), 'seqcap' (for whole exome sequencing), 'seqcap_probe' (64 Mb SeqCap EZ Probe v3), or 'wgs' (for whole genome sequencing)"
    :swagger/default     "seqcap"
    :reason              "Only support msk, seqcap, seqcap_propbe or wgs"}))

(s/def ::do_assign
  (st/spec
   {:spec                boolean?
    :type                :bool
    :description         "A boolean for whether a cutoff should be applied to determine the final decision or just the features should be returned"
    :swagger/default     true
    :reason              "Only support true or false"}))

(s/def ::do_mva
  (st/spec
   {:spec                boolean?
    :type                :bool
    :description         "A boolean for whether multivariate analysis should be run"
    :swagger/default     true
    :reason              "Only support true or false"}))

(s/def ::lite_format
  (st/spec
   {:spec                boolean?
    :type                :bool
    :description         "Saves the output in a lite format when set to true"
    :swagger/default     true
    :reason              "Only support true or false"}))

(s/def ::check_msi
  (st/spec
   {:spec                boolean?
    :type                :bool
    :description         "A boolean which determines whether the user wants to identify micro-sattelite instable tumors"
    :swagger/default     true
    :reason              "Only support true or false"}))

(s/def ::filepath
  (st/spec
   {:spec                (s/and string? #(re-matches #"^[a-zA-Z0-9]+:\/\/(\/|\.\/)[a-zA-Z0-9_]+.*" %))
    :type                :string
    :description         "File path for covertor."
    :swagger/default     nil
    :reason              "The filepath must be string."}))

(s/def ::parameters
  (s/keys :req-un []
          :opt-un [::filetype ::tumor_type ::data ::do_assign ::do_mva ::lite_format ::check_msi]))

(def sigma-params-body
  "A spec for the body parameters."
  (s/keys :req-un [::filepath ::parameters]))

;;; ------------------------------------------------ Event Metadata -------------------------------------------------
(def metadata
  {:route ["/tool/sigma"
           {:tags ["Tool"]
            :post {:summary "Mutational signature analysis for low statistics SNV data."
                   :parameters {:body sigma-params-body}
                   :responses {201 {:body {:output_file string? :log_url string?}}}
                   :handler (fn [{{{:keys [filepath parameters]} :body} :parameters}]
                              (let [workdir (get-workdir)
                                    from-path (if (re-matches #"^file:\/\/\/.*" filepath)
                                                ; Absolute path with file://
                                                (str/replace filepath #"^file:\/\/" "")
                                                (fs-lib/join-paths workdir (str/replace filepath #"^file:\/\/" "")))
                                    uuid (u/uuid)
                                    relative-dir (fs-lib/join-paths "download" uuid)
                                    output-file (fs-lib/join-paths relative-dir "sigma-results.csv")
                                    log-path (fs-lib/join-paths relative-dir "log")
                                    to-dir (fs-lib/join-paths workdir relative-dir)]
                                (fs-lib/create-directories! to-dir)
                                ; Launch the sigma
                                (spit (fs-lib/join-paths workdir log-path) (json/write-str {:status "Running" :msg ""}))
                                (events/publish-event! :sigma-convert
                                                       {:input-file from-path
                                                        :parameters parameters
                                                        :dest-dir to-dir})
                                {:status 201
                                 :body {:output_file output-file
                                        :log_url log-path}}))}
            :get {:summary "A json shema for sigma."
                  :parameters {}
                  :responses {200 {:body map?}}
                  :handler (fn [_]
                             {:status 200
                              :body (json-schema/transform sigma-params-body)})}}]
   :manifest {:description "Mutational signature analysis for low statistics SNV data."
              :category "Tool"
              :home "https://github.com/clinico-omics/tservice-plugins"
              :name "Mutational Signature Analysis"
              :source "PGx"
              :short_name "sigma"
              :icons [{:src "", :type "image/png", :sizes "192x192"}
                      {:src "", :type "image/png", :sizes "192x192"}]
              :author "Parklab"
              :hidden false
              :id "98b3aa7d24b6bdd4a4eeef5c9a117164"
              :app_name "parklab/sigma"}})

(def ^:const sigma-topics
  "The `Set` of event topics which are subscribed to for use in sigma tracking."
  #{:sigma-convert})

(def ^:private sigma-channel
  "Channel for receiving event sigma we want to subscribe to for sigma events."
  (async/chan))

;;; ------------------------------------------------ Event Processing ------------------------------------------------

(defn- sigma!
  [input-file dest-dir parameters]
  (let [log-path (fs-lib/join-paths dest-dir "log")
        output-file (fs-lib/join-paths dest-dir "sigma-results.csv")]
    (try
      (log/info (format "Mutational signature analysis for low statistics SNV data: [%s %s %s]" input-file parameters output-file))
      (let [sigma-result (sigma-lib/call-sigma! input-file output-file parameters)
            log (json/write-str sigma-result)]
        (log/info "Status: " sigma-result)
        (spit log-path log))
      (catch Exception e
        (let [log (json/write-str {:status "Error" :msg (.toString e)})]
          (log/info "Status: " log)
          (spit log-path log))))))

(defn- process-sigma-event!
  "Handle processing for a single event notification received on the sigma-channel"
  [sigma-event]
  ;; try/catch here to prevent individual topic processing exceptions from bubbling up.  better to handle them here.
  (try
    (when-let [{topic :topic object :item} sigma-event]
      ;; TODO: only if the definition changed??
      (case (events/topic->model topic)
        "sigma" (sigma! (:input-file object) (:dest-dir object) (:parameters object))))
    (catch Throwable e
      (log/warn (format "Failed to process sigma event. %s" (:topic sigma-event)) e))))

;;; --------------------------------------------------- Lifecycle ----------------------------------------------------

(defn events-init
  "Automatically called during startup; start event listener for sigma events."
  []
  (events/start-event-listener! sigma-topics sigma-channel process-sigma-event!))
